src/chainstate/stacks/boot/pox_2_tests.rs

```diff
@@ -0,0 +1,322 @@
+        .block_height_to_reward_cycle(burnchain.pox_constants.v1_unlock_height as u64)
+        .unwrap()
+        + 1;
+
+    eprintln!("First v2 cycle = {}", first_v2_cycle);
+
+    let epochs = StacksEpoch::all(0, 25 + 10);
+
+    let (mut peer, mut keys) = instantiate_pox_peer_with_epoch(
+        &burnchain,
+        "test_simple_pox_lockup_transition_pox_2",
+        6002,
+        Some(epochs.clone()),
+    );
+
+    let num_blocks = 35;
+
+    let alice = keys.pop().unwrap();
+    let bob = keys.pop().unwrap();
+    let charlie = keys.pop().unwrap();
+
+    let mut alice_reward_cycle = 0;
+
+    for tenure_id in 0u32..num_blocks {
+        let microblock_privkey = StacksPrivateKey::new();
+        let microblock_pubkeyhash =
+            Hash160::from_node_public_key(&StacksPublicKey::from_private(&microblock_privkey));
+        let tip = SortitionDB::get_canonical_burn_chain_tip(&peer.sortdb.as_ref().unwrap().conn())
+            .unwrap();
+
+        let (burn_ops, stacks_block, microblocks) = peer.make_tenure(
+            |ref mut miner,
+             ref mut sortdb,
+             ref mut chainstate,
+             vrf_proof,
+             ref parent_opt,
+             ref parent_microblock_header_opt| {
+                let parent_tip = get_parent_tip(parent_opt, chainstate, sortdb);
+                let coinbase_tx = make_coinbase(miner, tenure_id as usize);
+
+                let mut block_txs = vec![coinbase_tx];
+
+                if tenure_id == 1 {
+                    // Alice locks up exactly 25% of the liquid STX supply, so this should succeed.
+                    // this locks for cycles 6, 7, 8, 9.
+                    //  however, the v2 unlock occurs before cycle 8, so alice will be unlocked before cycles 8 and 9
+                    let alice_lockup = make_pox_lockup(
+                        &alice,
+                        0,
+                        1024 * POX_THRESHOLD_STEPS_USTX,
+                        AddressHashMode::SerializeP2PKH,
+                        key_to_stacks_addr(&alice).bytes,
+                        4,
+                        tip.block_height,
+                    );
+                    block_txs.push(alice_lockup);
+                } else if tenure_id == 10 {
+                    // Lets have Bob lock up for v2
+                    // this will lock for cycles 8, 9, 10, and 11
+                    //  the first v2 cycle will be 8
+                    let bob_lockup = make_pox_2_lockup(
+                        &bob,
+                        0,
+                        512 * POX_THRESHOLD_STEPS_USTX,
+                        AddressHashMode::SerializeP2PKH,
+                        key_to_stacks_addr(&bob).bytes,
+                        6,
+                        tip.block_height,
+                    );
+                    block_txs.push(bob_lockup);
+                } else if tenure_id == 11 {
+                    // Bob tries to lock in PoX v1 too, but it shouldn't work!
```

Comment:
I'm not sure this is checked below?  We're checking the same address in the reward set either way, so if either `make_pox_2_lockup()` or `make_pox_lockup()` succeed, the test will pass.

Reply:
Also, how would this error manifest if Bob called into `.pox` when 2.1 rolls around?  Would the block itself be unmineable because it calls into a contract that is no longer "active," or is this just a runtime error in the transaction, or is Bob's call re-routed to the `.pox-2` contract automatically?

Reply:
As currently implemented, it's a runtime error when someone attempts to lock via PoX after PoX 2 becomes active. If you look at the logs from this test, you'll see the tx being mined and aborted via a runtime error. But really, I should add a TestObserver to be able to check tx receipts in the test.

Reply:
(This is, I think, the correct behavior in the event that we _don't_ re-route pox calls to pox-2)

Reply:
Yes, that sounds right -- we don't do contract-call interception for pox.

But my other point stands -- it's not clear from what is being tested which contract-call is succeeding.  I don't think you need to go so far as to use a TestObserver (unless you want to truly confirm that the tx is getting rejected at runtime) -- I think you can get away with having Bob use one BTC address in pox, and use another in pox-2, and see which one got accepted.

Reply:
Hmmm, adding a TestObserver is better I think -- it allows for more direct/clear assertions (checking the PoX set is somewhat indirect for this, and there's other reasons Bob might not end up being included in the PoX1 reward set) and also richer assertions. It'll probably be useful for other PRs as well.